File: tests/python/test_reliability.py
import pytest
import numpy as np
from semiconductor_simulator import Simulator

def test_reliability_simulation():
    simulator = Simulator()
    simulator.initialize_geometry(10, 10)
    mask = np.zeros((10, 10))
    mask[2:5, 2:5] = 0.1
    simulator.run_exposure(13.5, 0.33, mask.tolist())
    simulator.run_metallization(0.5, "Cu", "pvd")
    simulator.run_deposition(0.1, "SiO2", "cvd")
    simulator.perform_electrical_test("resistance")
    simulator.simulate_thermal(300.0, 0.001)
    simulator.perform_reliability_test(0.001, 5.0)
    
    wafer = simulator.get_wafer()
    mttf = wafer.get_electromigration_mttf()
    stress = wafer.get_thermal_stress()
    field = wafer.get_dielectric_field()
    
    assert mttf.max() > 0, "MTTF should have non-zero values"
    assert mttf[mttf > 0].min() > 1e6, "MTTF should be > 1e6 s"
    assert stress.max() > 0, "Stress should have non-zero values"
    assert stress.max() < 1000, "Stress should be < 1000 MPa"
    assert field.max() > 0, "Field should have non-zero values"
    assert field.max() < 1e7, "Field should be < 10 MV/cm"

def test_reliability_gui():
    from PySide6.QtWidgets import QApplication
    from semiconductor_simulator.gui.reliability_panel import ReliabilityPanel
    app = QApplication([])
    
    panel = ReliabilityPanel()
    panel.current_input.setText("0.001")
    panel.voltage_input.setText("5.0")
    
    # Mock wafer setup
    panel.simulator.initialize_geometry(10, 10)
    mask = np.zeros((10, 10))
    mask[2:5, 2:5] = 0.1
    panel.simulator.run_exposure(13.5, 0.33, mask.tolist())
    panel.simulator.run_metallization(0.5, "Cu", "pvd")
    panel.simulator.run_deposition(0.1, "SiO2", "cvd")
    panel.simulator.perform_electrical_test("resistance")
    panel.simulator.simulate_thermal(300.0, 0.001)
    
    panel.run_test()
    assert "Min MTTF" in panel.result_label.text(), "Result should include MTTF"
    assert "Max Stress" in panel.result_label.text(), "Result should include stress"
    assert "Max Field" in panel.result_label.text(), "Result should include field"
    assert panel.error_label.text() == "", "No errors should occur"
    
    # Test metric selector
    panel.metric_selector.setCurrentText("Electromigration")
    assert panel.metric_selector.currentText() == "Electromigration"
    panel.metric_selector.setCurrentText("Thermal Stress")
    assert panel.metric_selector.currentText() == "Thermal Stress"
    panel.metric_selector.setCurrentText("Dielectric Breakdown")
    assert panel.metric_selector.currentText() == "Dielectric Breakdown"
    panel.metric_selector.setCurrentText("None")
    assert panel.metric_selector.currentText() == "None"

File: tests/cpp/test_reliability.cpp
#include "../../src/cpp/core/wafer.hpp"
#include "../../src/cpp/modules/reliability/reliability_model.hpp"
#include "../../src/cpp/modules/metallization/metallization_model.hpp"
#include "../../src/cpp/modules/thermal/thermal_model.hpp"
#include "../../src/cpp/modules/packaging/packaging_model.hpp"
#include <memory>
#include <cassert>
#include <iostream>

void test_reliability() {
    // Initialize wafer
    auto wafer = std::make_shared<Wafer>(300.0, 775.0, "silicon");
    wafer->initializeGrid(10, 10);

    // Add metal layer
    MetallizationModel metallization;
    Eigen::ArrayXXd photoresist = Eigen::ArrayXXd::Zero(10, 10);
    photoresist.block(2, 2, 3, 3).setConstant(0.1); // Metal pad
    wafer->setPhotoresistPattern(photoresist);
    metallization.simulateMetallization(wafer, 0.5, "Cu", "pvd");

    // Perform electrical test
    PackagingModel packaging;
    packaging.performElectricalTest(wafer, "resistance");

    // Run thermal simulation
    ThermalSimulationModel thermal;
    thermal.simulateThermal(wafer, 300.0, 0.001); // 300 K, 1 mA

    // Run reliability test
    ReliabilityModel reliability;
    reliability.performReliabilityTest(wafer, 0.001, 5.0); // 1 mA, 5 V

    // Check electromigration MTTF
    auto mttf = wafer->getElectromigrationMTTF();
    bool has_mttf = false;
    for (int i = 0; i < mttf.rows(); ++i) {
        for (int j = 0; j < mttf.cols(); ++j) {
            if (mttf(i, j) > 0) {
                has_mttf = true;
                assert(mttf(i, j) > 1e6); // MTTF > 1e6 s
            }
        }
    }
    assert(has_mttf);

    // Check thermal stress
    auto stress = wafer->getThermalStress();
    bool has_stress = false;
    for (int i = 0; i < stress.rows(); ++i) {
        for (int j = 0; j < stress.cols(); ++j) {
            if (stress(i, j) > 0) {
                has_stress = true;
                assert(stress(i, j) < 1000); // Stress < 1000 MPa
            }
        }
    }
    assert(has_stress);

    // Check dielectric field
    wafer->addFilmLayer(0.1, "SiO2");
    reliability.performReliabilityTest(wafer, 0.001, 5.0);
    auto field = wafer->getDielectricField();
    bool has_field = false;
    for (int i = 0; i < field.rows(); ++i) {
        for (int j = 0; j < field.cols(); ++j) {
            if (field(i, j) > 0) {
                has_field = true;
                assert(field(i, j) < 1e7); // Field < 10 MV/cm
            }
        }
    }
    assert(has_field);

    std::cout << "Reliability test passed\n";
}

int main() {
    test_reliability();
    return 0;
}